src/sources/audiosourceflac.cpp

```diff
@@ -0,0 +1,373 @@
+    if (m_file.isSequential()) {
+        return false;
+    }
+    return m_file.atEnd();
+}
+
+FLAC__StreamDecoderWriteStatus AudioSourceFLAC::flacWrite(
+        const FLAC__Frame *frame, const FLAC__int32 * const buffer[]) {
+    // decode buffer must be empty before decoding the next frame
+    DEBUG_ASSERT(m_decodeSampleBufferReadOffset >= m_decodeSampleBufferWriteOffset);
+    // reset decode buffer
+    m_decodeSampleBufferReadOffset = 0;
+    m_decodeSampleBufferWriteOffset = 0;
+    if (getChannelCount() != frame->header.channels) {
+        qWarning() << "Corrupt or unsupported FLAC file:"
+                << "Invalid number of channels in FLAC frame header"
+                << frame->header.channels << "<>" << getChannelCount();
+        return FLAC__STREAM_DECODER_WRITE_STATUS_ABORT;
+    }
+    if (getFrameRate() != frame->header.sample_rate) {
+        qWarning() << "Corrupt or unsupported FLAC file:"
+                << "Invalid sample rate in FLAC frame header"
+                << frame->header.sample_rate << "<>" << getFrameRate();
+        return FLAC__STREAM_DECODER_WRITE_STATUS_ABORT;
+    }
+    const SampleBuffer::size_type maxBlocksize = samples2frames(
+            m_decodeSampleBuffer.size());
+    if (maxBlocksize < frame->header.blocksize) {
+        qWarning() << "Corrupt or unsupported FLAC file:"
+                << "Block size in FLAC frame header exceeds the maximum block size"
+                << frame->header.blocksize << ">" << maxBlocksize;
+    }
+    switch (getChannelCount()) {
+    case 1: {
+        // optimized code for 1 channel (mono)
+        DEBUG_ASSERT(1 <= frame->header.channels);
+        for (unsigned i = 0; i < frame->header.blocksize; ++i) {
+            m_decodeSampleBuffer[m_decodeSampleBufferWriteOffset++] =
+                    buffer[0][i] * m_sampleScale;
+        }
+        break;
+    }
+    case 2: {
+        // optimized code for 2 channels (stereo)
+        DEBUG_ASSERT(2 <= frame->header.channels);
+        for (unsigned i = 0; i < frame->header.blocksize; ++i) {
+            m_decodeSampleBuffer[m_decodeSampleBufferWriteOffset++] =
+                    buffer[0][i] * m_sampleScale;
+            m_decodeSampleBuffer[m_decodeSampleBufferWriteOffset++] =
+                    buffer[1][i] * m_sampleScale;
+        }
+        break;
+    }
+    default: {
+        // generic code for multiple channels
+        DEBUG_ASSERT(getChannelCount() == frame->header.channels);
+        for (unsigned i = 0; i < frame->header.blocksize; ++i) {
+            for (unsigned j = 0; j < frame->header.channels; ++j) {
+                m_decodeSampleBuffer[m_decodeSampleBufferWriteOffset++] =
+                        buffer[j][i] * m_sampleScale;
+            }
+        }
+    }
+    } DEBUG_ASSERT(m_decodeSampleBufferReadOffset <= m_decodeSampleBufferWriteOffset);
+    return FLAC__STREAM_DECODER_WRITE_STATUS_CONTINUE;
+}
+
+void AudioSourceFLAC::flacMetadata(const FLAC__StreamMetadata *metadata) {
```

Comment:
According to 
https://xiph.org/flac/api/group__flac__stream__decoder.html#ga9d6df4a39892c05955122cf7f987f856
it is not clear if this will be called within a stream from FLAC__stream_decoder_process_single
Would this we save?